crates/web-plugins/didcomm-messaging/protocol-plugins/handler/src/handler.rs

```diff
@@ -0,0 +1,394 @@
+                    "found duplicate entry in plugin registry: {}",
+                    plugin.name()
+                );
+                return Err(PluginContainerError::DuplicateEntry);
+            }
+        }
+
+        // Reset collection of routes
+        self.collected_routes.clear();
+
+        // Mount plugins and collect routes on successful status
+        let errors: HashMap<_, _> = self
+            .plugins
+            .iter()
+            .filter_map(|plugin| {
+                let mut plugin = plugin.lock().unwrap();
+                match plugin.mount() {
+                    Ok(_) => {
+                        tracing::info!("mounted plugin {}", plugin.name());
+                        self.collected_routes.insert(plugin.name().to_string(), plugin.routes()); 
+                        None
+                    }
+                    Err(err) => {
+                        tracing::error!("error mounting plugin {}", plugin.name());
+                        Some((plugin.name().to_string(), err))
+                    }
+                }
+            })
+            .collect();
+
+        // Flag as loaded
+        self.loaded = true;
+
+        // Return state of completion
+        if errors.is_empty() {
+            tracing::debug!("plugin container loaded");
+            Ok(())
+        } else {
+            Err(PluginContainerError::PluginErrorMap(errors))
+        }
+    }
+
+    /// Unload container plugins
+    pub fn unload(&mut self) -> Result<(), PluginContainerError> {
+        // Unmount plugins and clear the HashMap of routes
+        let errors: HashMap<_, _> = self
+            .plugins
+            .iter()
+            .filter_map(|plugin| {
+                let plugin = plugin.lock().unwrap();
+                match plugin.unmount() {
+                    Ok(_) => {
+                        tracing::info!("unmounted plugin {}", plugin.name());
+                        None
+                    }
+                    Err(err) => {
+                        tracing::error!("error unmounting plugin {}", plugin.name());
+                        Some((plugin.name().to_owned(), err))
+                    }
+                }
+            })
+            .collect();
+
+        // Flag as unloaded
+        self.loaded = false;
+
+        // Return state of completion
+        if errors.is_empty() {
+            self.collected_routes.clear();
+            tracing::debug!("plugin container unloaded");
+            Ok(())
+        } else {
+            Err(PluginContainerError::PluginErrorMap(errors))
+        }
+    }
+
+    /// Return the route associated with a given name.
+    pub fn get_route(&self, name: &str) -> Result<Router, PluginContainerError> {
+        if self.loaded {
+            self.collected_routes
+                .get(name)
+                .cloned()
+                .ok_or(PluginContainerError::Unloaded)
+        } else {
+            Err(PluginContainerError::Unloaded)
+        }
+    }
```

Comment:
please can you give more comments to this function i think i understand what you are trying to do but not very sure, and you have provided no test for it too